File: src/io/AntiHome.java
package io;

import annotation.Draft;
import annotation.Intention;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Per-user "anti" home layout, created once so the engine always has a stable
 * place to write before any subsystem touches the disk. Lives in {@code io} so
 * an installer can invoke {@link #ensure()} as part of setup.
 *
 *   ~/anti/
 *     projects/    - user project workspaces
 *     logs/        - engine binary logs (io.Log default sink)
 *     placeholder/ - reserved scratch area
 */
@Draft
@Intention("Bootstrap the ~/anti directory tree once so Log, projects, and future bakery tooling have a canonical home before they run. Installer-facing via ensure().")
public final class AntiHome {

    private static final String ROOT_NAME = "anti";
    private static final String PROJECTS_NAME = "projects";
    private static final String LOGS_NAME = "logs";
    private static final String PLACEHOLDER_NAME = "placeholder";

    private static volatile boolean ensured;

    private AntiHome() {
    }

    public static String root() {
        return dir("");
    }

    public static String projects() {
        return dir(PROJECTS_NAME);
    }

    public static String logs() {
        return dir(LOGS_NAME);
    }

    public static String placeholder() {
        return dir(PLACEHOLDER_NAME);
    }

    private static String dir(String sub) {
        return Path.of(System.getProperty("user.home"), ROOT_NAME, sub).toString();
    }

    /** Creates the full layout; idempotent. Returns true when every dir exists. */
    public static boolean ensure() {
        if (ensured) {
            return true;
        }
        try {
            Files.createDirectories(Path.of(root()));
            Files.createDirectories(Path.of(projects()));
            Files.createDirectories(Path.of(logs()));
            Files.createDirectories(Path.of(placeholder()));
            ensured = true;
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    /** Default log file: ~/anti/logs/engine.bin (truncated every run by FileWriter). */
    public static String defaultLogPath() {
        ensure();
        return Path.of(logs(), "engine.bin").toString();
    }
}

File: src/io/antihome.c
#include "io/antihome.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "io/file.h"

// antihome.c — AntiHome port (Legacy: io/AntiHome.java). ~/anti layout.

#define ROOT_NAME      "anti"
#define PROJECTS_NAME  "projects"
#define LOGS_NAME      "logs"
#define PLACEHOLDER    "placeholder"
#define LOG_FILE       "engine.bin"

static char root_buf[FILE_PATH_MAX];
static char projects_buf[FILE_PATH_MAX];
static char logs_buf[FILE_PATH_MAX];
static char placeholder_buf[FILE_PATH_MAX];
static bool ensured;

static void build_path(char *out, const char *sub) {
    const char *home = getenv("HOME");
    if (!home || *home == '\0')
        home = ".";
    if (*sub == '\0')
        snprintf(out, FILE_PATH_MAX, "%s/%s", home, ROOT_NAME);
    else
        snprintf(out, FILE_PATH_MAX, "%s/%s/%s", home, ROOT_NAME, sub);
}

const char *AntiHome_root(void) {
    build_path(root_buf, "");
    return root_buf;
}

const char *AntiHome_projects(void) {
    build_path(projects_buf, PROJECTS_NAME);
    return projects_buf;
}

const char *AntiHome_logs(void) {
    build_path(logs_buf, LOGS_NAME);
    return logs_buf;
}

const char *AntiHome_placeholder(void) {
    build_path(placeholder_buf, PLACEHOLDER);
    return placeholder_buf;
}

bool AntiHome_ensure(void) {
    if (ensured)
        return true;
    if (!File_mkdirs(AntiHome_root()))
        return false;
    if (!File_mkdirs(AntiHome_projects()))
        return false;
    if (!File_mkdirs(AntiHome_logs()))
        return false;
    if (!File_mkdirs(AntiHome_placeholder()))
        return false;
    ensured = true;
    return true;
}

const char *AntiHome_defaultLogPath(void) {
    AntiHome_ensure();
    snprintf(logs_buf, FILE_PATH_MAX, "%s/%s", AntiHome_logs(), LOG_FILE);
    return logs_buf;
}